[Lyft-GA] Enable color consistency in a dashboard

khtruong opened the pull request:
Accidentally messed up rebasing and it seemed easier to just create a new branch.

Refer to this PR for all details: https://github.com/apache/incubator-superset/pull/7086

khtruong committed:
Enable color consistency in a dashboard

Moved actions, minor UI, allowed dashboard copy

Fix linting errors

Undo unintentional change

Updated and added unit tests

Fail quietly if package has not been updated

Fail quietly on dashboard copy if package is old

khtruong committed:
Update packages

khtruong committed:
Remove unnecessary code

graceguo-supercat commented on superset/assets/src/dashboard/actions/dashboardState.js:
```diff
@@ -261,6 +261,18 @@ export function removeSliceFromDashboard(id) {
+export function setColorSchemeAndUnsavedChanges(colorScheme) {
+  return dispatch => {
+    dispatch(setColorScheme(colorScheme));
+    dispatch(setUnsavedChanges(true));
```
setUnsavedChanges (and `hasUnsavedChanges` state) is shared with dashboard's layout change. If you change color scheme,  change dashboard layout, then undo a all dashboard layout change, the hasUnsavedChanges flag will be reset. 
you probably have to add some logic to make sure this flag carries changes from both.

khtruong replied:
I added a bool flag to track unsaved color changes. I don't love the approach but since we want the layout stuff to be refactored to allow non-layout changes to be undo-able, this was the least amount of changes until that refactoring gets done.

graceguo-supercat commented on superset/assets/src/dashboard/components/ColorComponentPane.jsx:
```diff
@@ -0,0 +1,107 @@
+            onMouseEnter={this.onMouseEnter}
+            onMouseLeave={this.onMouseLeave}
+          >
+            <ColorSchemeControl
```
can you add scheme name as tooltip when user mouse over color scheme choices?

khtruong replied:
Let me get @dorq 's opinion. The color name schemes may not make sense to users based on what we call them today.

dorq replied:
@khtruong and I discussed this offline, and we agreed that even though the color scheme names are not very user-friendly (e.g. 'lyftColors', 'bnbColors', 'd3Category10', d3Category20'), there are situations in which it would be valuable to see them.  So adding them as a tooltip (as opposed to putting the scheme names directly into the dropdown options) is a good compromise.

@khtruong will add this to the PR (thanks, Kim!)

khtruong replied:
I'll try to make them as user-friendly as possible (i.e. lyftColors -> Lyft Colors, bnbColors -> Airbnb Colors, etc)

graceguo-supercat replied:
i don't really mean to use scheme name (like lyftcolors bnbcolors), any user friendly descriptive name will be good. so that when i want to apply same color scheme all over multiple dashboards, it will be easier for me to remember.

khtruong replied:
<img width="409" alt="Screen Shot 2019-03-28 at 10 23 29 AM" src="https://user-images.githubusercontent.com/47833996/55180714-62957280-5147-11e9-95d5-121f1ffd1ab4.png">

<img width="412" alt="Screen Shot 2019-03-28 at 10 23 43 AM" src="https://user-images.githubusercontent.com/47833996/55180718-645f3600-5147-11e9-996e-2319ae201435.png">

I left scheme names as is unless someone has better names.

dorq replied:
LGTM

kristw replied:
`ColorScheme` datastructure has `label` field.
I think the code in this PR should use the `label` instead of the `keys` which are `id`.
Then we can update the color schemes in `@superset-ui` to include human-friendly `label`.

p.s. `ColorScheme` also has `description` field.

kristw replied:
![image](https://user-images.githubusercontent.com/1659771/55181281-9b821700-5148-11e9-900e-9a729c9d71bd.png)

khtruong replied:
I just updated the color schemes in @superset-ui to have actual labels.

khtruong replied:
https://github.com/apache-superset/superset-ui/pull/125

kristw replied:
awesome!

khtruong committed:
Addressed Grace's comments

khtruong committed:
Small fix for item key

graceguo-supercat commented:
Please see my screenshot. Color scheme is carried from dashboard to explore view, but color is not consistent for each label:

![j9UBYSSIyG](https://user-images.githubusercontent.com/27990562/55186747-81e6cc80-5154-11e9-838c-1f025d09c81e.gif)

khtruong commented:
Yeah. Unfortunately, the chart and dashboard coloring are completely separate. I chatted with design and PM on this. The issue is that charts can be part of multiple dashboards so which dashboard should carry over to the chart? So the color scheme is reset when you explore the chart. I think it appeared the color scheme carried over but I would guess that coincidentally the dashboard and chart both have the same color scheme. We haven't thought of a great to get a consistent experience.

graceguo-supercat commented:
> Yeah. Unfortunately, the chart and dashboard coloring are completely separate. I chatted with design and PM on this. The issue is that charts can be part of multiple dashboards so which dashboard should carry over to the chart? So the color scheme is reset when you explore the chart. I think it appeared the color scheme carried over but I would guess that coincidentally the dashboard and chart both have the same color scheme. We haven't thought of a great to get a consistent experience.

it's not coincident, because you added `colorScheme` param into charts' formData, and explore view used same formData which is updated from dashboard.

kristw commented on superset/assets/package.json:
```diff
@@ -49,7 +49,7 @@
   "dependencies": {
     "@data-ui/sparkline": "^0.0.54",
     "@superset-ui/chart": "^0.9.6",
```
Please check versions. `master` has newer version for these `@superset-ui/xxx`

khtruong replied:
This is a branch off of lyftga. I think lyftga is a little behind.

kristw replied:
oh....haha my bad

khtruong commented:
> > Yeah. Unfortunately, the chart and dashboard coloring are completely separate. I chatted with design and PM on this. The issue is that charts can be part of multiple dashboards so which dashboard should carry over to the chart? So the color scheme is reset when you explore the chart. I think it appeared the color scheme carried over but I would guess that coincidentally the dashboard and chart both have the same color scheme. We haven't thought of a great to get a consistent experience.
> 
> it's not coincident, because you added `colorScheme` param into charts' formData, and `Explore chart` action will open the chart with formData which is updated from dashboard. It's different from open chart by id from explore view, `Explore chart` will carry additional filter parameters from dashboard too.
> 
> In my opinion, chart opened from dashboard should be consistent with its look/data inside the dashboard.

Eek. You're right. I was thinking of when I changed the colors and then went to the chart and explored it from the chart list. I have updated it to use the chart form data so that when you explore it, it will use the color scheme associated with that chart.

Yeah, I think there are differing opinions of what should happen. I personally am unsure and think we need to rethink about it from a broader perspective that covers all, or at least most, cases. It does feel more consistent to have what the dashboard set but we still have a problem if the chart is part of multiple dashboards that have different color schemes. Should the color scheme just be temporary? Should we indicate to the user that this is not the chart's saved color scheme? There hasn't been a general consensus yet. We can definitely discuss more for future PRs if you want to brainstorm a better experience. But for now, I will reset it to the chart's color scheme until there's been a clearer consensus of that experience should be.

khtruong committed:
Reset chart's color during exploration

graceguo-supercat commented:
> It does feel more consistent to have what the dashboard set but we still have a problem if the chart is part of multiple dashboards that have different color schemes.

Think about this: 
you have a chart `Population` show world population, and you insert it into multiple dashboards. in dashboard_1 you have dashboard level filter say country=US, and dashboard_2 you have filter say country=UK. 
- if you open chart by itself, no filter, the chart show population for all over the world.
- The chart should show US population in dashboard_1.
- The chart should show UK population in dashboard_2.
- if you `Explore chart` from dashboard_1, you should see chart show same number as it in dashboard_1.
My opinion is, color scheme, should behave same as this `country` parameter. 

Chart `for sure` will be used in multiple dashboards. If we open explore view from `Explore chart` button of a dashboard, the chart should show same color and same data as it was in that specific dashboard.

khtruong committed:
Do not reset chart form data when exploring chart

khtruong commented:
> > It does feel more consistent to have what the dashboard set but we still have a problem if the chart is part of multiple dashboards that have different color schemes.
> 
> Think about this:
> you have a chart `Population` show world population, and you insert it into multiple dashboards. in dashboard_1 you have dashboard level filter say country=US, and dashboard_2 you have filter say country=UK.
> 
> * if you open chart by itself, no filter, the chart show population for all over the world.
> * The same chart should show US population in dashboard_1.
> * The same chart should show UK population in dashboard_2.
> * if you click `Explore chart` button from dashboard_1, you should see chart show same number as it in dashboard_1.
>   My opinion is, color scheme, should behave same as this `country` parameter.
> 
> Chart `for sure` will be used in multiple dashboards. If we open explore view from `Explore chart` button of a dashboard, the chart should show `same data`, as well as `same label color`, it was in that specific dashboard.

Good point! Grace and I discussed offline. We are in agreement on how the ideal scenario should be. But I think that leads to the next point: we need color consistency in charts as well. I'd like to address these in separate PR's that build on top of each other where each PR gets us to a better experience (It's always nicer to review smaller PRs if we can help it. I know I appreciate that :) ). So I'd like to keep it how it is and open up a git issue that will address color consistency in charts that includes having consistency when you explore charts from the dashboard. How's that sound?

graceguo-supercat commented:
good. We can take care dashboard => explore view label consistency in another PR. Thanks for the work!!

khtruong commented:
> good. We can take care dashboard => explore view label consistency in another PR. Thanks for the work!!

Thanks for all the feedback!! You have been awesome.

khtruong commented:
> good. We can take care dashboard => explore view label consistency in another PR. Thanks for the work!!

Issue has been created to track this work: https://github.com/apache/incubator-superset/issues/7163
